sinks/kafka/driver.go

```
@@ -0,0 +1,161 @@
+	}
+	message := &proto.Message{Value: []byte(string(jsonItems))}
+	_, err = self.producer.Produce(topic, partition, message)
+	if err != nil {
+		return fmt.Errorf("failed to produce message to %s:%d: %s", topic, partition, err)
+	}
+	return nil
+}
+
+func (self *kafkaSink) DebugInfo() string {
+	info := fmt.Sprintf("%s\n", self.Name())
+	info += fmt.Sprintf("There are two kafka's topics: %s,%s:\n", self.eventsTopic, self.timeSeriesTopic)
+	info += fmt.Sprintf("The kafka's broker list is: %s", self.sinkBrokerHosts)
+	return info
+}
+
+func (self *kafkaSink) Name() string {
+	return "Apache-Kafka Sink"
+}
+
+func init() {
+	extpoints.SinkFactories.Register(NewKafkaSink, "kafka")
+}
+
+func NewKafkaSink(uri *url.URL, _ extpoints.HeapsterConf) ([]sink_api.ExternalSink, error) {
+	var kafkaSink kafkaSink
+	opts, err := url.ParseQuery(uri.RawQuery)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parser url's query string: %s", err)
+	}
+
+	if len(opts["timeseriestopic"]) < 1 {
+		return nil, fmt.Errorf("There is no timeseriestopic assign for config kafka-sink")
+	}
+	kafkaSink.timeSeriesTopic = opts["eventstopic"][0]
+
+	if len(opts["eventstopic"]) < 1 {
+		return nil, fmt.Errorf("There is no eventstopic assign for config kafka-sink")
+	}
+	kafkaSink.eventsTopic = opts["eventstopic"][0]
+
+	if len(opts["brokers"]) < 1 {
+		return nil, fmt.Errorf("There is no broker assign for connecting kafka broker")
+	}
+	kafkaSink.sinkBrokerHosts = append(kafkaSink.sinkBrokerHosts, opts["brokers"]...)
+
+	//connect to kafka cluster
+	brokerConf := kafka.NewBrokerConf(brokerClientID)
+	brokerConf.DialTimeout = brokerDialTimeout
+	brokerConf.DialRetryLimit = brokerDialRetryLimit
+	brokerConf.DialRetryWait = brokerDialRetryWait
+	brokerConf.LeaderRetryLimit = brokerLeaderRetryLimit
+	brokerConf.LeaderRetryWait = brokerLeaderRetryWait
+	brokerConf.AllowTopicCreation = true
+
+	broker, err := kafka.Dial(kafkaSink.sinkBrokerHosts, brokerConf)
```

You can take a look at the influxdb sink which retries instead of failing hard.